clients/application/src/main/java/org/eclipse/hono/client/application/DownstreamMessage.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (c) 2021 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+
+package org.eclipse.hono.client.application;
+
+import java.util.Objects;
+
+import io.vertx.core.buffer.Buffer;
+
+/**
+ * A message of Hono's northbound APIs, flowing from the messaging system to the backend application.
+ */
+public abstract class DownstreamMessage implements Message {
```

Comment:
FMPOV this should be refactored into an interface and an `AbstractDownstreamMessage` implementation ...